### PR TITLE
Optimize complexity of part deduplication in parallel replicas announcement

### PR DESCRIPTION
**Pull request by davenger**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Previously the algorithmic complexity of part deduplication logic in parallel replica announcement handling was O(n^2) which could take noticeable time for tables with many part (or partitions). This change makes the complexity O(n*log(n)) 

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit 2650a2062899f5f232176ee56814cce66c800139 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/69596/2650a2062899f5f232176ee56814cce66c800139/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/2fce90ab76b851198b39041c9e7c4b03ad601ff7/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Comment by davenger** [comment]
Stress test (debug) : https://github.com/ClickHouse/ClickHouse/issues/69573

File: src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp [hunk 1]
```diff
@@ -882,12 +881,28 @@ void InOrderCoordinator<mode>::doHandleInitialAllRangesAnnouncement(InitialAllRa
...
+                continue;
+
+            /// Also look at the previous part, it could be covering the current one
+            if (covering_it != all_parts_to_read.begin())
+            {
+                --covering_it;
+                if (is_covered_or_covering(*covering_it))
+                    continue;
+            }
+        }
```
Comment (nickitat) [inline src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp:905]: looks correct, but i'd implement a sanity check for intersecting parts in the resulting `all_parts_to_read`. under `#ifdef DEBUG` probably.

**Comment by nickitat** [comment]
it seems default coordinator has the same problem: https://github.com/ClickHouse/ClickHouse/blob/17c1e82bc0daaf6fcc51f6db198371e1464bb16d/src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp#L379-L390

File: src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp [hunk 2]
```diff
@@ -911,6 +911,24 @@ void InOrderCoordinator<mode>::doHandleInitialAllRangesAnnouncement(InitialAllRa
         std::sort(ranges.begin(), ranges.end());
     }
 
+#ifndef NDEBUG
+    /// Double check that there are no intersecting parts
+    {
+        auto part_it = all_parts_to_read.begin();
```
Comment (nickitat) [inline src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp:920]: maybe `adjacent_find` would look more accurate

File: src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp [hunk 3]
```diff
@@ -376,17 +376,20 @@ void DefaultCoordinator::initializeReadingState(InitialAllRangesAnnouncement ann
...
+        for (auto && part : announcement.description)
+        {
+            auto intersecting_it = known_parts.lower_bound(Part{.description = part, .replicas = {}});
+
+            if (intersecting_it != known_parts.end() && !intersecting_it->description.info.isDisjoint(part.info))
+                throw Exception(ErrorCodes::LOGICAL_ERROR, "Intersecting parts found in announcement");
 
-        all_parts_to_read.push_back(Part{.description = std::move(part), .replicas = {announcement.replica_num}});
+            all_parts_to_read.push_back(Part{.description = std::move(part), .replicas = {announcement.replica_num}});
+            known_parts.emplace(Part{.description = part, .replicas = {}});
```
Comment (nickitat) [inline src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp:391]: `part` was already moved from

**Review by nickitat — APPROVED** [review]

**Pull request merged**